File: fons/norma/hal/codegen/ts/solum.ts
/**
 * solum.ts - File System Implementation
 *
 * Native TypeScript implementation of the HAL solum (filesystem) interface.
 * Uses Bun file APIs and Node.js fs/path/os modules.
 */

import * as fs from 'node:fs/promises';
import * as fsSync from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';

export const solum = {
    // =========================================================================
    // READING
    // =========================================================================

    async lege(filePath: string): Promise<string> {
        return Bun.file(filePath).text();
    },

    async legeOctetos(filePath: string): Promise<Uint8Array> {
        return new Uint8Array(await Bun.file(filePath).arrayBuffer());
    },

    async *legens(filePath: string): AsyncIterable<Uint8Array> {
        const stream = Bun.file(filePath).stream();
        for await (const chunk of stream) {
            yield chunk;
        }
    },

    // =========================================================================
    // WRITING
    // =========================================================================

    async scribe(filePath: string, data: string): Promise<void> {
        await Bun.write(filePath, data);
    },

    async scribeOctetos(filePath: string, data: Uint8Array): Promise<void> {
        await Bun.write(filePath, data);
    },

    async appone(filePath: string, data: string): Promise<void> {
        await fs.appendFile(filePath, data);
    },

    async *scribens(filePath: string): AsyncGenerator<void, void, Uint8Array> {
        const handle = await fs.open(filePath, 'w');
        try {
            while (true) {
                const chunk = yield;
                if (chunk === undefined) break;
                await handle.write(chunk);
            }
        }
        finally {
            await handle.close();
        }
    },

    // =========================================================================
    // FILE INFO
    // =========================================================================

    exstat(filePath: string): boolean {
        try {
            fsSync.accessSync(filePath);
            return true;
        }
        catch {
            return false;
        }
    },

    estLimae(filePath: string): boolean {
        try {
            return fsSync.statSync(filePath).isFile();
        }
        catch {
            return false;
        }
    },

    estDirectorii(filePath: string): boolean {
        try {
            return fsSync.statSync(filePath).isDirectory();
        }
        catch {
            return false;
        }
    },

    async magnitudo(filePath: string): Promise<number> {
        const stats = await fs.stat(filePath);
        return stats.size;
    },

    async modificatum(filePath: string): Promise<number> {
        const stats = await fs.stat(filePath);
        return stats.mtimeMs;
    },

    // =========================================================================
    // FILE OPERATIONS
    // =========================================================================

    async dele(filePath: string): Promise<void> {
        await fs.unlink(filePath);
    },

    async copia(src: string, dest: string): Promise<void> {
        await fs.copyFile(src, dest);
    },

    async move(src: string, dest: string): Promise<void> {
        await fs.rename(src, dest);
    },

    async tange(filePath: string): Promise<void> {
        const now = new Date();
        try {
            await fs.utimes(filePath, now, now);
        }
        catch {
            // File doesn't exist, create it
            await Bun.write(filePath, '');
        }
    },

    // =========================================================================
    // DIRECTORY OPERATIONS
    // =========================================================================

    async creaDir(dirPath: string): Promise<void> {
        await fs.mkdir(dirPath, { recursive: true });
    },

    async elenca(dirPath: string): Promise<string[]> {
        return fs.readdir(dirPath);
    },

    async *ambula(dirPath: string): AsyncIterable<string> {
        const entries = await fs.readdir(dirPath, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(dirPath, entry.name);
            if (entry.isDirectory()) {
                yield* solum.ambula(fullPath);
            }
            else {
                yield fullPath;
            }
        }
    },

    async deleDir(dirPath: string): Promise<void> {
        await fs.rmdir(dirPath);
    },

    async deleArborem(dirPath: string): Promise<void> {
        await fs.rm(dirPath, { recursive: true, force: true });
    },

    // =========================================================================
    // PATH UTILITIES
    // =========================================================================

    iunge(parts: string[]): string {
        return path.join(...parts);
    },

    dir(filePath: string): string {
        return path.dirname(filePath);
    },

    basis(filePath: string): string {
        return path.basename(filePath);
    },

    extensio(filePath: string): string {
        return path.extname(filePath);
    },

    absolve(filePath: string): string {
        return path.resolve(filePath);
    },

    domus(): string {
        return os.homedir();
    },

    temp(): string {
        return os.tmpdir();
    },
};

File: fons/nanus-go/subsidia/norma/hal/solum.go
// solum.go - File System Implementation (Go)
//
// Native Go implementation of the HAL solum (filesystem) interface.
// Uses os, io, and path/filepath packages.

package hal

import (
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"time"
)

// SolumStatus holds file status information
type SolumStatus struct {
	Modus        int64
	Nexus        int64
	Possessor    int64
	Grex         int64
	Magnitudo    int64
	Modificatum  int64
	EstDirectorii bool
	EstVinculum  bool
}

// Solum provides filesystem operations
var Solum = struct {
	// Reading
	Lege        func(path string) string
	LegeOctetos func(path string) []byte

	// Writing
	Scribe        func(path string, data string)
	ScribeOctetos func(path string, data []byte)
	Appone        func(path string, data string)

	// File Info
	Exstat        func(path string) bool
	EstLimae      func(path string) bool
	EstDirectorii func(path string) bool
	Magnitudo     func(path string) int64
	Modificatum   func(path string) int64

	// File Status
	Status       func(path string) SolumStatus
	LegeVinculum func(path string) string

	// File Operations
	Dele  func(path string)
	Copia func(src, dest string)
	Move  func(src, dest string)
	Tange func(path string)

	// Directory Operations
	CreaDir     func(path string)
	Elenca      func(path string) []string
	DeleDir     func(path string)
	DeleArborem func(path string)

	// Path Utilities
	Iunge    func(parts []string) string
	Dir      func(path string) string
	Basis    func(path string) string
	Extensio func(path string) string
	Absolve  func(path string) string
	Domus    func() string
	Temp     func() string
}{
	// =========================================================================
	// READING
	// =========================================================================

	Lege: func(path string) string {
		data, err := os.ReadFile(path)
		if err != nil {
			panic(err)
		}
		return string(data)
	},

	LegeOctetos: func(path string) []byte {
		data, err := os.ReadFile(path)
		if err != nil {
			panic(err)
		}
		return data
	},

	// =========================================================================
	// WRITING
	// =========================================================================

	Scribe: func(path string, data string) {
		err := os.WriteFile(path, []byte(data), 0644)
		if err != nil {
			panic(err)
		}
	},

	ScribeOctetos: func(path string, data []byte) {
		err := os.WriteFile(path, data, 0644)
		if err != nil {
			panic(err)
		}
	},

	Appone: func(path string, data string) {
		f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			panic(err)
		}
		defer f.Close()
		if _, err := f.WriteString(data); err != nil {
			panic(err)
		}
	},

	// =========================================================================
	// FILE INFO
	// =========================================================================

	Exstat: func(path string) bool {
		_, err := os.Stat(path)
		return err == nil
	},

	EstLimae: func(path string) bool {
		info, err := os.Stat(path)
		if err != nil {
			return false
		}
		return info.Mode().IsRegular()
	},

	EstDirectorii: func(path string) bool {
		info, err := os.Stat(path)
		if err != nil {
			return false
		}
		return info.IsDir()
	},

	Magnitudo: func(path string) int64 {
		info, err := os.Stat(path)
		if err != nil {
			panic(err)
		}
		return info.Size()
	},

	Modificatum: func(path string) int64 {
		info, err := os.Stat(path)
		if err != nil {
			panic(err)
		}
		return info.ModTime().UnixMilli()
	},

	// =========================================================================
	// FILE STATUS
	// =========================================================================

	Status: func(path string) SolumStatus {
		info, err := os.Lstat(path)
		if err != nil {
			panic(err)
		}
		return SolumStatus{
			Modus:        int64(info.Mode().Perm()),
			Nexus:        1, // Go doesn't expose nlink portably
			Possessor:    0, // Would need syscall for uid
			Grex:         0, // Would need syscall for gid
			Magnitudo:    info.Size(),
			Modificatum:  info.ModTime().UnixMilli(),
			EstDirectorii: info.IsDir(),
			EstVinculum:  info.Mode()&fs.ModeSymlink != 0,
		}
	},

	LegeVinculum: func(path string) string {
		target, err := os.Readlink(path)
		if err != nil {
			panic(err)
		}
		return target
	},

	// =========================================================================
	// FILE OPERATIONS
	// =========================================================================

	Dele: func(path string) {
		if err := os.Remove(path); err != nil {
			panic(err)
		}
	},

	Copia: func(src, dest string) {
		srcFile, err := os.Open(src)
		if err != nil {
			panic(err)
		}
		defer srcFile.Close()

		destFile, err := os.Create(dest)
		if err != nil {
			panic(err)
		}
		defer destFile.Close()

		if _, err := io.Copy(destFile, srcFile); err != nil {
			panic(err)
		}
	},

	Move: func(src, dest string) {
		if err := os.Rename(src, dest); err != nil {
			panic(err)
		}
	},

	Tange: func(path string) {
		now := time.Now()
		if err := os.Chtimes(path, now, now); err != nil {
			// File doesn't exist, create it
			f, err := os.Create(path)
			if err != nil {
				panic(err)
			}
			f.Close()
		}
	},

	// =========================================================================
	// DIRECTORY OPERATIONS
	// =========================================================================

	CreaDir: func(path string) {
		if err := os.MkdirAll(path, 0755); err != nil {
			panic(err)
		}
	},

	Elenca: func(path string) []string {
		entries, err := os.ReadDir(path)
		if err != nil {
			panic(err)
		}
		names := make([]string, len(entries))
		for i, entry := range entries {
			names[i] = entry.Name()
		}
		return names
	},

	DeleDir: func(path string) {
		if err := os.Remove(path); err != nil {
			panic(err)
		}
	},

	DeleArborem: func(path string) {
		if err := os.RemoveAll(path); err != nil {
			panic(err)
		}
	},

	// =========================================================================
	// PATH UTILITIES
	// =========================================================================

	Iunge: func(parts []string) string {
		return filepath.Join(parts...)
	},

	Dir: func(path string) string {
		return filepath.Dir(path)
	},

	Basis: func(path string) string {
		return filepath.Base(path)
	},

	Extensio: func(path string) string {
		return filepath.Ext(path)
	},

	Absolve: func(path string) string {
		abs, err := filepath.Abs(path)
		if err != nil {
			panic(err)
		}
		return abs
	},

	Domus: func() string {
		home, err := os.UserHomeDir()
		if err != nil {
			panic(err)
		}
		return home
	},

	Temp: func() string {
		return os.TempDir()
	},
}
